### PR TITLE
Allow more non-inline modules in blocks

### LINKED ISSUE #36772: non-line mod without path attribute false positive
This code

```Rust
fn main() {
    #[path="foo"]
    mod foo {
        mod bar;
    }
}
```

Fails to compile because there is no path on the `bar`, however I think it should work because there is a path on the `foo`. That is, the following is valid and will look for `foo/bar.rs`:

```Rust
fn main() {}

#[path="foo"]
mod foo {
    mod bar;
}
```

cc @jseyfried https://github.com/rust-lang/rust/commit/d21e908175de9c70fffc9c07d676557392cf2b39

### PR DESCRIPTION
**Pull request by jseyfried**
Currently, non-inline modules without a `#[path]` attribute are not allowed in blocks.
This PR allows non-inline modules that have an ancestor module with a `#[path]` attribute, provided there is not a nearer ancestor block.

For example,

``` rust
fn main() {
    #[path = "..."] mod foo {
        mod bar; //< allowed by this PR
        fn f() {
            mod bar; //< still an error
        }
    }
}
```

Fixes #36772.
r? @nikomatsakis

### CODE REVIEW

**Comment by jseyfried** [comment]
c.f. #31534
cc @matklad

File: src/libsyntax/ext/base.rs [hunk 1]
```diff
@@ -696,7 +696,7 @@ pub struct ExpansionData {
     pub depth: usize,
     pub backtrace: ExpnId,
     pub module: Rc<ModuleData>,
-    pub in_block: bool,
+    pub no_noninline_mod: bool,
```
Comment (nikomatsakis) [inline src/libsyntax/ext/base.rs]: I think this field could use a comment. The meaning of its name is not immediately obvious on its own, I'd say...

**Review by nikomatsakis — APPROVED** [review]
r=me with a comment on the field

File: src/libsyntax/ext/base.rs (hunk 1 above)
Comment (jseyfried) [inline src/libsyntax/ext/base.rs]: Good point.

**Pull request merged**